driver-core/src/main/java/com/datastax/driver/core/RequestHandler.java

```diff
@@ -692,13 +731,22 @@ public void onException(Connection connection, Exception exception, long latency
             Host queriedHost = current;
             try {
                 connection.release();
-
                 if (exception instanceof ConnectionException) {
-                    if (metricsEnabled())
-                        metrics().getErrorMetrics().getConnectionErrors().inc();
                     ConnectionException ce = (ConnectionException)exception;
-                    logError(ce.address, ce);
-                    retry(false, null);
+                    RetryPolicy retryPolicy = retryPolicy();
+                    RetryPolicy.RetryDecision decision;
+                    if (retryPolicy instanceof ExtendedRetryPolicy)
+                        decision = ((ExtendedRetryPolicy)retryPolicy).onConnectionError(statement, request().consistency(), ce, retriesByPolicy);
+                    else
+                        decision = RetryPolicy.RetryDecision.tryNextHost(null);
+                    if (metricsEnabled()) {
+                        metrics().getErrorMetrics().getConnectionErrors().inc();
+                        if (decision.getType() == Type.RETRY)
+                            metrics().getErrorMetrics().getRetriesOnConnectionError().inc();
+                        if (decision.getType() == Type.IGNORE)
+                            metrics().getErrorMetrics().getIgnoresOnConnectionError().inc();
+                    }
+                    processRetryDecision(decision, connection, exception);
                     return;
                 }
                 setFinalException(connection, exception);
```

Comment:
Same remark as in onSet: if it's a completely unexpected exception, it sounds reasonable to fail fast and report it back to the user through the future.